syne_tune/blackbox_repository/blackbox_surrogate.py

```diff
@@ -135,70 +240,199 @@ def make_model_pipeline(configuration_space, fidelity_space, model):
+            else:
+                # when no fidelity is given and a fidelity space exists, we
+                # return all fidelities
+                # we construct a input dataframe with all fidelity for the
+                # configuration given to call the transformer at once which
+                # is more efficient due to vectorization
+                surrogate_input_df = pd.DataFrame(
+                    [surrogate_input] * self.num_fidelities
+                )
+                surrogate_input_df[fidelity_attr] = self.fidelity_values
+                prediction = self._transform_from_finite_differences(
+                    self.surrogate_pipeline[seed].predict(surrogate_input_df)
+                )
+            extract_fidelity = do_fit_diffs and single_fidelity
+        else:
+            # Multivariate regression
+            prediction = self.surrogate_pipeline[seed].predict(
                 pd.DataFrame([surrogate_input])
             )
+            prediction = self._transform_from_finite_differences(
+                prediction.reshape((self.num_fidelities, -1))
+            )
+            extract_fidelity = single_fidelity
 
-            # converts the returned nd-array with shape (1, num_metrics) to the list of objectives values
-            prediction = prediction.reshape(-1).tolist()
+        if extract_fidelity:
+            assert self.fidelity_values is not None, "blackbox has no fidelities"
+            # If there are several fidelity values, pick the first
+            fidelity = list(fidelity.values())[0]
+            ind = np.where(self.fidelity_values == fidelity)
+            assert ind, f"fidelity {fidelity} not among {self.fidelity_values}"
+            ind = ind[0]
+            prediction = dict(zip(self.objectives_names, prediction[ind]))
+        return prediction
 
-            # convert prediction to dictionary
-            return dict(zip(self.objectives_names, prediction))
-        else:
-            # when no fidelity is given and a fidelity space exists, we return all fidelities
-            # we construct a input dataframe with all fidelity for the configuration given to call the transformer
-            # at once which is more efficient due to vectorization
-            surrogate_input_df = pd.DataFrame(
-                [surrogate_input] * len(self.fidelity_values)
-            )
-            surrogate_input_df[
-                next(iter(self.fidelity_space.keys()))
-            ] = self.fidelity_values
-            objectives_values = self.surrogate_pipeline.predict(surrogate_input_df)
-            return objectives_values
+    def hyperparameter_objectives_values(
+        self, predict_curves: bool = False
+    ) -> Tuple[pd.DataFrame, pd.DataFrame]:
+        raise NotImplementedError("This is a surrogate already!")
 
 
-def add_surrogate(blackbox: Blackbox, surrogate=None, configuration_space=None):
+def add_surrogate(
+    blackbox: Blackbox,
+    surrogate=None,
```

Comment:
typing?

Reply:
You mean for `surrogate`? This is just a bit complicated, some sci-kit learn type.